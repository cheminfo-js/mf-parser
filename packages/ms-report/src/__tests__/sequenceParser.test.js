'use strict';

const sequenceParser = require('../sequenceParser');

describe('sequenceParser', () => {
  it('AA(Ph)A(Ph)A(Ts)A(Ph)U', () => {
    let result = sequenceParser('AA(Ph)A(Ph)A(Ts)A(Ph)U');
    expect(result).toStrictEqual({
      begin: 'H',
      end: 'OH',
      parts: [
        { value: 'Ala', code: 'A' },
        { value: 'Ala(Ph)', code: 'α' },
        { value: 'Ala(Ph)', code: 'α' },
        { value: 'Ala(Ts)', code: '¹α' },
        { value: 'Ala(Ph)', code: 'α' },
        { value: 'Sec', code: 'Θ' }
      ],
      alternatives: { α: { count: 2 } },
      replacements: {
        'Ala(Ph)': { code: 'α', residue: 'Ala', modification: 'Ph' },
        'Ala(Ts)': { code: '¹α', residue: 'Ala', modification: 'Ts' },
        Sec: { code: 'Θ', id: 'Sec' }
      }
    });
  });

  it('HAlaAlaAlaOH', () => {
    let result = sequenceParser('HAlaAlaAlaOH');
    expect(result).toStrictEqual({
      alternatives: {},
      begin: 'H',
      end: 'OH',
      parts: [
        { code: 'A', value: 'Ala' },
        { code: 'A', value: 'Ala' },
        { code: 'A', value: 'Ala' }
      ],
      replacements: {}
    });
  });

  it('HAlaAla(H-1OH)AlaOH', () => {
    let result = sequenceParser('HAlaAla(H-1OH)AlaOH');
    expect(result).toStrictEqual({
      alternatives: { α: { count: 1 } },
      begin: 'H',
      end: 'OH',
      parts: [
        { code: 'A', value: 'Ala' },
        { code: 'α', value: 'Ala(H-1OH)' },
        { code: 'A', value: 'Ala' }
      ],
      replacements: {
        'Ala(H-1OH)': { code: 'α', modification: 'H-1OH', residue: 'Ala' }
      }
    });
  });

  it('H(+)AlaAla(H-1OH)AlaOH', () => {
    let result = sequenceParser('H(+)AlaAla(H-1OH)AlaOH');
    expect(result).toStrictEqual({
      alternatives: { α: { count: 1 } },
      begin: 'H(+)',
      end: 'OH',
      parts: [
        { code: 'A', value: 'Ala' },
        { code: 'α', value: 'Ala(H-1OH)' },
        { code: 'A', value: 'Ala' }
      ],
      replacements: {
        'Ala(H-1OH)': { code: 'α', modification: 'H-1OH', residue: 'Ala' }
      }
    });
  });

  it('ForAlaAla(H-1OH)AlaOH', () => {
    let result = sequenceParser('(For)AlaAla(H-1OH)AlaOH');
    expect(result).toStrictEqual({
      alternatives: { α: { count: 1 } },
      begin: 'For',
      end: 'OH',
      parts: [
        { value: 'Ala', code: 'A' },
        { value: 'Ala(H-1OH)', code: 'α' },
        { value: 'Ala', code: 'A' }
      ],
      replacements: {
        'Ala(H-1OH)': { code: 'α', residue: 'Ala', modification: 'H-1OH' }
      }
    });
  });

  it('(MeO)AA(H-1NH2)AA rna', () => {
    let result = sequenceParser('(MeO)AA(H-1NH2)AA', { kind: 'rna' });
    expect(result).toStrictEqual({
      begin: 'MeO',
      end: 'H',
      parts: [
        { value: 'Amp', code: 'A' },
        { value: 'Amp(H-1NH2)', code: 'α' },
        { value: 'Amp', code: 'A' },
        { value: 'Amp', code: 'A' }
      ],
      alternatives: { α: { count: 1 } },
      replacements: {
        'Amp(H-1NH2)': { code: 'α', residue: 'Amp', modification: 'H-1NH2' }
      }
    });
  });
});
