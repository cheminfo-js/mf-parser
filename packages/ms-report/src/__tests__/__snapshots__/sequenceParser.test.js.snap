// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sequenceParser (MeO)AA(H-1NH2)AA rna 1`] = `
Object {
  "all": Array [
    Object {
      "kind": "begin",
      "label": "MeO",
    },
    Object {
      "fromBegin": 1,
      "fromEnd": 4,
      "kind": "residue",
      "label": "A",
      "value": "Amp",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 3,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Amp(H-1NH2)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 2,
      "kind": "residue",
      "label": "A",
      "value": "Amp",
    },
    Object {
      "fromBegin": 4,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Amp",
    },
    Object {
      "kind": "end",
      "label": "H",
    },
  ],
  "alternatives": Object {
    "α": Object {
      "count": 1,
    },
  },
  "begin": Object {
    "kind": "begin",
    "label": "MeO",
  },
  "end": Object {
    "kind": "end",
    "label": "H",
  },
  "replacements": Object {
    "Amp(H-1NH2)": Object {
      "label": "α",
      "modification": "H-1NH2",
      "residue": "Amp",
    },
  },
  "residues": Array [
    Object {
      "fromBegin": 1,
      "fromEnd": 4,
      "kind": "residue",
      "label": "A",
      "value": "Amp",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 3,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Amp(H-1NH2)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 2,
      "kind": "residue",
      "label": "A",
      "value": "Amp",
    },
    Object {
      "fromBegin": 4,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Amp",
    },
  ],
}
`;

exports[`sequenceParser AA(Ph)A(Ph)A(Ts)A(Ph)U 1`] = `
Object {
  "all": Array [
    Object {
      "kind": "begin",
      "label": "H",
    },
    Object {
      "fromBegin": 1,
      "fromEnd": 6,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 5,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(Ph)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 4,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(Ph)",
    },
    Object {
      "fromBegin": 4,
      "fromEnd": 3,
      "kind": "residue",
      "label": "¹α",
      "replaced": true,
      "value": "Ala(Ts)",
    },
    Object {
      "fromBegin": 5,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(Ph)",
    },
    Object {
      "fromBegin": 6,
      "fromEnd": 1,
      "kind": "residue",
      "label": "Θ",
      "replaced": true,
      "value": "Sec",
    },
    Object {
      "kind": "end",
      "label": "OH",
    },
  ],
  "alternatives": Object {
    "α": Object {
      "count": 2,
    },
  },
  "begin": Object {
    "kind": "begin",
    "label": "H",
  },
  "end": Object {
    "kind": "end",
    "label": "OH",
  },
  "replacements": Object {
    "Ala(Ph)": Object {
      "label": "α",
      "modification": "Ph",
      "residue": "Ala",
    },
    "Ala(Ts)": Object {
      "label": "¹α",
      "modification": "Ts",
      "residue": "Ala",
    },
    "Sec": Object {
      "id": "Sec",
      "label": "Θ",
    },
  },
  "residues": Array [
    Object {
      "fromBegin": 1,
      "fromEnd": 6,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 5,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(Ph)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 4,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(Ph)",
    },
    Object {
      "fromBegin": 4,
      "fromEnd": 3,
      "kind": "residue",
      "label": "¹α",
      "replaced": true,
      "value": "Ala(Ts)",
    },
    Object {
      "fromBegin": 5,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(Ph)",
    },
    Object {
      "fromBegin": 6,
      "fromEnd": 1,
      "kind": "residue",
      "label": "Θ",
      "replaced": true,
      "value": "Sec",
    },
  ],
}
`;

exports[`sequenceParser ForAlaAla(H-1OH)AlaOH 1`] = `
Object {
  "all": Array [
    Object {
      "kind": "begin",
      "label": "For",
    },
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(H-1OH)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "kind": "end",
      "label": "OH",
    },
  ],
  "alternatives": Object {
    "α": Object {
      "count": 1,
    },
  },
  "begin": Object {
    "kind": "begin",
    "label": "For",
  },
  "end": Object {
    "kind": "end",
    "label": "OH",
  },
  "replacements": Object {
    "Ala(H-1OH)": Object {
      "label": "α",
      "modification": "H-1OH",
      "residue": "Ala",
    },
  },
  "residues": Array [
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(H-1OH)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
  ],
}
`;

exports[`sequenceParser H(+)AlaAla(H-1OH)AlaOH 1`] = `
Object {
  "all": Array [
    Object {
      "kind": "begin",
      "label": "H(+)",
    },
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(H-1OH)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "kind": "end",
      "label": "OH",
    },
  ],
  "alternatives": Object {
    "α": Object {
      "count": 1,
    },
  },
  "begin": Object {
    "kind": "begin",
    "label": "H(+)",
  },
  "end": Object {
    "kind": "end",
    "label": "OH",
  },
  "replacements": Object {
    "Ala(H-1OH)": Object {
      "label": "α",
      "modification": "H-1OH",
      "residue": "Ala",
    },
  },
  "residues": Array [
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(H-1OH)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
  ],
}
`;

exports[`sequenceParser HAlaAla(H-1OH)AlaOH 1`] = `
Object {
  "all": Array [
    Object {
      "kind": "begin",
      "label": "H",
    },
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(H-1OH)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "kind": "end",
      "label": "OH",
    },
  ],
  "alternatives": Object {
    "α": Object {
      "count": 1,
    },
  },
  "begin": Object {
    "kind": "begin",
    "label": "H",
  },
  "end": Object {
    "kind": "end",
    "label": "OH",
  },
  "replacements": Object {
    "Ala(H-1OH)": Object {
      "label": "α",
      "modification": "H-1OH",
      "residue": "Ala",
    },
  },
  "residues": Array [
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "α",
      "replaced": true,
      "value": "Ala(H-1OH)",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
  ],
}
`;

exports[`sequenceParser HAlaAlaAlaOH 1`] = `
Object {
  "all": Array [
    Object {
      "kind": "begin",
      "label": "H",
    },
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "kind": "end",
      "label": "OH",
    },
  ],
  "alternatives": Object {},
  "begin": Object {
    "kind": "begin",
    "label": "H",
  },
  "end": Object {
    "kind": "end",
    "label": "OH",
  },
  "replacements": Object {},
  "residues": Array [
    Object {
      "fromBegin": 1,
      "fromEnd": 3,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 2,
      "fromEnd": 2,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
    Object {
      "fromBegin": 3,
      "fromEnd": 1,
      "kind": "residue",
      "label": "A",
      "value": "Ala",
    },
  ],
}
`;
